### ISSUE TITLE
Scale down bug on AWS

### ISSUE THREAD

**Issue by sharpevo** [kubernetes/autoscaler#144]
There's 3 nodes in an Autoscaling Group of AWS:
- Node-A, utilization: 0.3, kube-system pods present, "Protect From Scale In".
- Node-B, utilization: 0.0, may be removed
- Node-C, utilization: 0.8

Cluster Autoscaler reports that the `Node-B` will be removed in a given `scale-down-delay`, but AWS removes `Node-A` due to the desired capacity shrinking request by user, rather than the `Node-B`.

```
At 2017-06-28T06:06:36Z instance i-0b9d was taken out of service 
in response to a user request, shrinking the capacity from 3 to 2.
```
`Node-A` has pods in namespace `kube-system`, so I guess the termination should be triggered by AWS , before the Cluster Autoscaler takes any actions. Does Cluster Autoscaler send requests to change desired capacity first? But there's already `ShouldDecrementDesiredCapacity` in [aws_manager.go](https://github.com/kubernetes/autoscaler/blob/28caf01651b517d3c0cd47de2bd5daed7e75e7ad/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L141)

kubernetes: v1.6.6
cluster-autoscaler: v0.5.4
autoscaling group:
```
{
    "AutoScalingGroups": [
        {
            "AutoScalingGroupName": "kube-test-asg",
            "AutoScalingGroupARN": "arn",
            "LaunchConfigurationName": "kube-test",
            "MinSize": 1,
            "MaxSize": 5,
            "DesiredCapacity": 2,
            "DefaultCooldown": 300,
            "LoadBalancerNames": [],
            "TargetGroupARNs": [],
            "HealthCheckType": "EC2",
            "HealthCheckGracePeriod": 300,
            "Instances": [
                {
                    "InstanceId": "i-0309",
                    "LifecycleState": "InService",
                    "HealthStatus": "Healthy",
                    "LaunchConfigurationName": "kube-test",
                    "ProtectedFromScaleIn": false
                },
                {
                    "InstanceId": "i-0af4",
                    "LifecycleState": "InService",
                    "HealthStatus": "Healthy",
                    "LaunchConfigurationName": "kube-test",
                    "ProtectedFromScaleIn": false
                }
            ],
            "CreatedTime": "2017-06-27T01:02:31.582Z",
            "SuspendedProcesses": [],
            "VPCZoneIdentifier": "subnet",
            "EnabledMetrics": [],
            "Tags": [],
            "TerminationPolicies": [
                "Default"
            ],
            "NewInstancesProtectedFromScaleIn": false
        }
    ]
}
```

**Comment by mwielgus** [comment]
While sending a request to AWS we explicitly specify which node should be taken down.
https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L139

https://docs.aws.amazon.com/sdk-for-go/api/service/autoscaling/#TerminateInstanceInAutoScalingGroupInput Suggests that it should work as expected. Maybe something wrong went on the AWS side? We had no prior reports about this type of issues.

@mumoshu @andrewsykim, any thoughts?

**Comment by mumoshu** [comment]
@mwielgus Thanks for checking in but unfortunately I'm not sure for now.

@sharpevo Does the issue consistently reproduce? Also, would you mind trying to reproduce it with older versions of CA?

**Comment by sharpevo** [comment]
@mwielgus @mumoshu ,

Previously, I've run the test twice, and always the other node been killed, with the same version of Kubernetes and CA. Running the test one more time today, I realized that the `Node-A` is probaly terminated by CA, for the reason of unregistered node.

CA try to remove node `i-045d079434eeca204` for the reason of unregistered:

```
I0629 09:42:14.568520       1 static_autoscaler.go:250] Calculating unneeded nodes
I0629 09:42:14.568531       1 scale_down.go:143] Node ip-172-31-26-231 - utilization 0.302672
I0629 09:42:14.568539       1 scale_down.go:143] Node ip-172-31-30-123 - utilization 0.000000
I0629 09:42:14.568548       1 cluster.go:74] Fast evaluation: ip-172-31-26-231 for removal
I0629 09:42:14.568593       1 cluster.go:88] Fast evaluation: node ip-172-31-26-231 cannot be removed: non-deamons set, non-mirrored, kube-system pod present: cluster-autoscaler-3893175531-5g8qj
I0629 09:42:14.568597       1 cluster.go:74] Fast evaluation: ip-172-31-30-123 for removal
I0629 09:42:14.568603       1 cluster.go:103] Fast evaluation: node ip-172-31-30-123 may be removed
I0629 09:42:14.568654       1 static_autoscaler.go:263] ip-172-31-30-123 is unneeded since 2017-06-29 09:42:14.568522453 +0000 UTC duration 127.716µs
I0629 09:42:14.568659       1 static_autoscaler.go:268] Starting scale down
I0629 09:42:14.568663       1 scale_down.go:198] ip-172-31-30-123 was unneeded for 138.511µs
I0629 09:42:14.568683       1 scale_down.go:237] No candidates for scale down
I0629 09:42:15.515786       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:17.031120       1 reflector.go:405] k8s.io/autoscaler/cluster-autoscaler/utils/kubernetes/listers.go:156: Watch close - *v1.Pod total 39 items received
I0629 09:42:17.519379       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:19.526666       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:21.536484       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:23.544884       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:24.651078       1 static_autoscaler.go:130] 2 unregistered nodes present
I0629 09:42:24.651087       1 utils.go:208] Removing unregistered node aws:///cn-north-1a/i-045d079434eeca204
W0629 09:42:24.671576       1 utils.go:220] Failed to remove node aws:///cn-north-1a/i-045d079434eeca204: min size reached, nodes will not be deleted
W0629 09:42:24.672328       1 static_autoscaler.go:137] Failed to remove unregistered nodes: min size reached, nodes will not be deleted
I0629 09:42:25.548522       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:27.556300       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:29.559792       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:31.574547       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
I0629 09:42:33.578231       1 leaderelection.go:204] succesfully renewed lease kube-system/cluster-autoscaler
```

The [UnregisteredNode](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/clusterstate/clusterstate.go#L90) is defined as:
>UnregisteredNode contains information about nodes that are present on the cluster provider side
>but failed to register in Kubernetes.

But it's the master & minion node in my kubernetes cluster, with PrivateIP `172.31.26.231` and InstanceId `i-045d079434eeca204` :

```
$ aws ec2 describe-instances

{
     "Reservations": [
         {
             "Groups": [],
             "Instances": [
                 {
                     "AmiLaunchIndex": 0,
                     "InstanceId": "i-045d079434eeca204",
                     "InstanceType": "t2.micro",
                     "LaunchTime": "2017-06-29T07:57:50.000Z",
                     "Monitoring": {
                         "State": "disabled"
                     },
                     "Placement": {
                         "AvailabilityZone": "cn-north-1a",
                         "GroupName": "",
                         "Tenancy": "default"
                     },
                     "PrivateIpAddress": "172.31.26.231",
                     "ProductCodes": [],
                     "State": {
                         "Code": 16,
                         "Name": "running"
                     },
  
...
}
```

and there's pods in `kube-system` namespace running in this node

```
[root@ip-172-31-26-231 cluster]# kubectl get pods -n kube-system -o wide
NAME                                  READY     STATUS    RESTARTS   AGE       IP           NODE
cluster-autoscaler-3893175531-5g8qj   1/1       Running   0          3h        172.17.0.2   ip-172-31-26-231
[root@ip-172-31-26-231 cluster]# kubectl get nodes
NAME               STATUS    AGE       VERSION
ip-172-31-26-231   Ready     4h        v1.6.6
[root@ip-172-31-26-231 cluster]# kubectl describe svc
Name:                   kubernetes
Namespace:              default
Labels:                 component=apiserver
                        provider=kubernetes
Annotations:            <none>
Selector:               <none>
Type:                   ClusterIP
IP:                     10.254.0.1
Port:                   https   443/TCP
Endpoints:              172.31.26.231:6443
Session Affinity:       ClientIP
Events:                 <none>

```

**Comment by mwielgus** [comment]
Please attach `kubectl get nodes -o yaml` of the unregistered node. There must be some discrepancies between provider id on the AWS side and in K8S.

**Comment by sharpevo** [comment]
@mwielgus Thank you!

Yes, there's not `ProviderId` in the node yaml, so that all the nodes are unregistered, according to https://github.com/kubernetes/autoscaler/blob/670b5b4b780c408296b2505fa3aa87f74887e83d/cluster-autoscaler/clusterstate/clusterstate.go#L725

I start the cluster manually, and am I missing the `--cloud-config` and `--cloud-provider` options in the `kube-apiserver`? Laterly, I append the following configs to `kube-apiserver`:

```
--cloud-provider=aws 
--cloud-config=/opt/cluster/config/cloud-config
```
but there's not any examples for aws at https://kubernetes.io/docs/getting-started-guides/scratch/#cloud-providers, so the `cloud-config` file only contains
```
[Global]
Zone = cn-north-1a
```

Additionally, I try to disable the unregistered nodes checking, and CA only provides `--unregistered-node-removal-time` options.


the node yaml:

```
[root@ip-172-31-26-231 cluster]# kubectl get nodes -o yaml
apiVersion: v1
items:
- apiVersion: v1
  kind: Node
  metadata:
    annotations:
      node.alpha.kubernetes.io/ttl: "0"
      volumes.kubernetes.io/controller-managed-attach-detach: "true"
    creationTimestamp: 2017-06-30T00:13:55Z
    labels:
      beta.kubernetes.io/arch: amd64
      beta.kubernetes.io/os: linux
      kubernetes.io/hostname: ip-172-31-26-231
    name: ip-172-31-26-231
    namespace: ""
    resourceVersion: "57"
    selfLink: /api/v1/nodesip-172-31-26-231
    uid: 01863e07-5d29-11e7-8a5f-0296c3480e94
  spec:
    externalID: ip-172-31-26-231
    providerID: aws:////i-045d079434eeca204
  status:
    addresses:
    - address: 172.31.26.231
      type: LegacyHostIP
    - address: 172.31.26.231
      type: InternalIP
    - address: ip-172-31-26-231
      type: Hostname
    allocatable:
      cpu: "1"
      memory: 912560Ki
      pods: "110"
    capacity:
      cpu: "1"
      memory: 1014960Ki
      pods: "110"
    conditions:
    - lastHeartbeatTime: 2017-06-30T00:14:05Z
      lastTransitionTime: 2017-06-30T00:13:55Z
      message: kubelet has sufficient disk space available
      reason: KubeletHasSufficientDisk
      status: "False"
      type: OutOfDisk
    - lastHeartbeatTime: 2017-06-30T00:14:05Z
      lastTransitionTime: 2017-06-30T00:13:55Z
      message: kubelet has sufficient memory available
      reason: KubeletHasSufficientMemory
      status: "False"
      type: MemoryPressure
    - lastHeartbeatTime: 2017-06-30T00:14:05Z
      lastTransitionTime: 2017-06-30T00:13:55Z
      message: kubelet has no disk pressure
      reason: KubeletHasNoDiskPressure
      status: "False"
      type: DiskPressure
    - lastHeartbeatTime: 2017-06-30T00:14:05Z
      lastTransitionTime: 2017-06-30T00:13:55Z
      message: kubelet is posting ready status
      reason: KubeletReady
      status: "True"
      type: Ready
    daemonEndpoints:
      kubeletEndpoint:
        Port: 10250
    images:
    - names:
      - gcr.io/google_containers/kubernetes-dashboard-amd64:v1.6.1
      sizeBytes: 134353887
    - names:
      - gcr.io/google_containers/cluster-autoscaler:v0.5.4
      sizeBytes: 124212102
    - names:
      - gcr.io/google_containers/heapster-amd64:v1.3.0
      sizeBytes: 68105973
    - names:
      - busybox@sha256:be3c11fdba7cfe299214e46edc642e09514dbb9bbefcd0d3836c05a1e0cd0642
      - busybox:latest
      sizeBytes: 1106304
    - names:
      - gcr.io/google-containers/pause-amd64:3.0
      sizeBytes: 746888
    nodeInfo:
      architecture: amd64
      bootID: 205d6e58-9c65-4bd6-9953-195bd002c2f7
      containerRuntimeVersion: docker://Unknown
      kernelVersion: 3.10.0-514.21.1.el7.x86_64
      kubeProxyVersion: v1.6.6
      kubeletVersion: v1.6.6
      machineID: ea4bb35ac7b55a78c3c4fb8cdb4b7f65
      operatingSystem: linux
      osImage: CentOS Linux 7 (Core)
      systemUUID: EC2CCBCA-1FF8-FB14-0915-7AA89AB53582
kind: List
metadata: {}
resourceVersion: ""
selfLink: ""
```

**Comment by sharpevo** [comment]
The reason may be the missing region or zone.

From the node yaml, the `ProviderID` looks like `aws:////i-045d079434eeca204`, then CA inserts the it into `registered` at https://github.com/kubernetes/autoscaler/blob/44e0897fa7e03a9ec4320d895bdd165fdb8b0609/cluster-autoscaler/clusterstate/clusterstate.go#L725
```
registered.Insert(node.Spec.ProviderID)
```

While executing `removeOldUnregisteredNodes`, the node name is `aws:///cn-north-1a/i-045d079434eeca204` it the CA log file, and it is different from the `ProviderID`, i.e., `aws:////i-045d079434eeca204`, that's why all the nodes are unregistered.

So what options am I missing in the kubernetes configs?

**Comment by sharpevo** [comment]
@mwielgus  @mumoshu 

The reason may be the missing `*instance.AvailabilityZone`.

From the node yaml, the `ProviderID` looks like `aws:////i-045d079434eeca204`, then CA inserts the it into `registered` map at https://github.com/kubernetes/autoscaler/blob/44e0897fa7e03a9ec4320d895bdd165fdb8b0609/cluster-autoscaler/clusterstate/clusterstate.go#L725
```
registered.Insert(node.Spec.ProviderID)
```

While executing `getNotRegisteredNodes`, the node from `nodeGroup.Nodes()`, a.k.a., `asg.awsManager.GetAsgNodes(asg)` in https://github.com/kubernetes/autoscaler/blob/6ce37d02d1340a49f331f914e0c5584fa1788ef1/cluster-autoscaler/cloudprovider/aws/aws_cloud_provider.go#L286, which returns the `node` formed with `*instance.AvailabilityZone`, https://github.com/kubernetes/autoscaler/blob/6ce37d02d1340a49f331f914e0c5584fa1788ef1/cluster-autoscaler/cloudprovider/aws/aws_manager.go#L320

```

	for _, instance := range group.Instances {
		result = append(result,
			fmt.Sprintf("aws:///%s/%s", *instance.AvailabilityZone, *instance.InstanceId))
	}
```

From the CA log file, it's `aws:///cn-north-1a/i-045d079434eeca204` , and is different from the `ProviderID`, i.e., `aws:////i-045d079434eeca204`, that's why all the nodes are unregistered.

So what options am I missing in the kubernetes configs?

**Comment by mumoshu** [comment]
@sharpevo AFAIK, kubelet seems to collect the provider id and push it to apiserver. Perhaps you're missing something in your kubelet setup - for example, are you sure you're providing `--cloud-provider=aws` not only apiserver but kubelet?

Relevant code in kubelet: https://github.com/kubernetes/kubernetes/blob/18e8bedb2bd7780ce9f1fd6ae9a427c0824b0d32/pkg/kubelet/kubelet_node_status.go#L270-L272

**Comment by sharpevo** [comment]
@mumoshu Thanks!

I missed the kubelet configs indeed! Addtionally, I found that a new action `ec2:DescribeInstances` should be added to the IAM policy, to avoid the error `error listing AWS instances: UnauthorizedOperation: You are not authorized to perform this operation.`

**Comment by mumoshu** [comment]
@sharpevo Great 👍 Thanks for reporting!

**Comment by mumoshu** [comment]
@sharpevo Regarding the `ec2:DescribeInstances` permission, do you have any insight on why it is needed? AFAIK, it isn't called explicitly in CA. Anyway, if it is needed, we would want to add the fact to [the doc](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/aws#permissions)

**Comment by sharpevo** [comment]
@mumoshu Sorry for taking a while to reply, I'm working on another project and not pay more attention on CA, because it works so well! Thank you for the great job :-)

About `ec2:DescribeInstances`, I've taken another test this morning, and find that it should be necesarry for the instances in ASG, or they can not connect with the master node, with the following errors from kubelet:
```
error: failed to run Kubelet: could not init cloud provider "aws": 
error finding instance i-077d207dc6fa9bfe5: error listing AWS instances: 
UnauthorizedOperation: You are not authorized to perform this operation.
        status code: 403, request id: c436d9fc-2109-4194-818c-f4e3d9ec0595
```

As a result, this policy works fine for me

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Stmt1498177749000",
            "Effect": "Allow",
            "Action": [
                "autoscaling:DescribeAutoScalingGroups",
                "autoscaling:DescribeAutoScalingInstances",
                "autoscaling:DescribeTags",
                "autoscaling:SetDesiredCapacity",
                "autoscaling:TerminateInstanceInAutoScalingGroup",
                "autoscaling:DescribeLaunchConfigurations",
                "ec2:DescribeInstances"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

BTW, CA may not scale up if there's some nodes terminated in aws but not deleted in kubernetes, so I have to launch CA with the option `--max-total-unready-percentage=100`.

**Comment by mumoshu** [comment]
@sharpevo Thanks for the information! It does make sense to me.

As far as I can say, `--max-total-unready-percentage=100` sounds like something we'd like to avoid, because otherwise CA will perhaps keep scaling down nodes even when there are considerable number of unready(=unfunctional for some reason) nodes.

Probably you would rather want to try automatically triggering `kubectl drain` on the terminating node so that nodes in aws and k8s could always be kept in sync?
I'm unsure which k8s cluster provisioning tool you're using but for example tools like kube-aws allows it.

**Comment by sharpevo** [comment]
@mumoshu I'm launching the cluster on bare metal, with the help of "systemd" to manage k8s components in master and minions. IMO, there're always some NotReady nodes in the cluster after  scaling up and down. If I understand CA correctly, the cluster will be "not ready for scaling" when the amount of NotReady nodes over the the `--ok-total-unready-count` or `--max-total-unready-percentage`. In order to keep the cluster healthy, I'm trying to define "not ready for scaling" with the meaning of "100% nodes are unready", rather than deleting NotReady nodes periodically. 

About the `kube-aws`, do you mean [nodeDrainer](https://github.com/kubernetes-incubator/kube-aws/blob/a1d127b1d2cb8ff44c3d3b56824dec0671a343d0/core/controlplane/config/templates/cluster.yaml#L1173)? `kubectl drains` will add a new status `SchedulingDisabled` to the nodes. It may be not so effective than deleting the NotReady nodes directly. Additionally, I've taken `--scale-down-unneeded-time` or `--scale-down-utilization-threshold` with an appropriate value, to prevent the cluster from scaling down some valuable nodes.

Here's my CA options:
```
--v=4
--log-dir=/var/log
--cloud-provider=aws
--skip-nodes-with-local-storage=true
--node-group-auto-discovery=asg:tag=ca-discovery
--scale-down-unneeded-time=20m
--scale-down-utilization-threshold=0.1
--unregistered-node-removal-time=20m
--max-total-unready-percentage=100
--kubernetes=http://172.31.31.253:8080
```

Forgiving my stupid, I still can not get the scenario of "unfunctional for some reason", if it's upgrading, why not modify "Launch Configuration" in ASG? A brand new AMI in the ASG might be a little bit more better for upgrading I think.

**Comment by mumoshu** [comment]
@sharpevo Thank you very much for the detailed explanation 👍 

> IMO, there're always some NotReady nodes in the cluster after scaling up and down.

Yes, I think so too.

> In order to keep the cluster healthy, I'm trying to define "not ready for scaling" with the meaning of "100% nodes are unready", rather than deleting NotReady nodes periodically.

Ah, I guess I now understand your intention! So, for example, you'd like CA to just add more nodes when there are insufficiency in cluster resource due to nodes that were once healthy but became unready afterwards, so that you can preserve enough healthy nodes even when there were extra unhealthy nodes. That makes sense to me.

Now, I'm now unsure why `--max-total-unready-percentage=100` isn't the default.
Hi @mwielgus, I'd appreciate it if you could clarify me the intended use-case of `--max-total-unready-percentage`.

> About the kube-aws, do you mean nodeDrainer?

Exactly 👍 

> kubectl drains will append a status SchedulingDisabled to the nodes. It seems not so effective than deleting the NotReady nodes directly.

I believe you are correct.
The node drainer would be useful only for cleanly terminating "Ready" nodes because what it does is, afaik, to mark a node SchedulingDisabled and then gracefully stop pods running on it. When the node is already "Unready", there could be no pods to be gracefully terminated by `kubectl drain` hence draining won't be useful.

**Comment by patrickshan** [comment]
@mumoshu @sharpevo sorry about replying this closed issue. But just want to mention a similar issue we found with kube v1.8.6 and cluster-autoscaler v1.0.0. We already have cloud provider set for both apiserver and kubelet. It seems that the issue was caused by AWS ASG "Instance Protection" setting which by default doesn't have "Protect From Scale In" turned on. so two activities inside the ASG happens at the same time (one by cluster-autoscaler to remove a specific node and the other by AWS ASG to reduce the size of the group) and both of them were trying to reduce the ASG size from N to N-1.
